File: LeetCode/53MaximumSubarry/MaximumSubarray.js
class MaxSubArray {
  constructor(list) {
    this.list = list
  }

  // https://leetcode.com/problems/maximum-subarray/discuss/139218/Javascript-very-clear-and-short-DP-solution
  maxSubArray1() {
    for (let i = 1; i < this.list.length; i++) {
      this.list[i] = Math.max(this.list[i], this.list[i] + this.list[i - 1])
    }
    console.log(this.list) // [ -2, 1, -2, 4, 3, 5, 6, 1, 5 ]
    return Math.max(...this.list)
  }

  // https://leetcode.com/problems/maximum-subarray/discuss/20193/DP-solution-and-some-thoughts
  maxSubArray2() {
    const n = this.list.length
    let dp = []
    dp[0] = this.list[0]
    let max = dp[0]
    for (let i = 1; i < n; i++) {
      dp[i] = this.list[i] + (dp[i - 1] > 0 ? dp[i - 1] : 0)
      max = Math.max(max, dp[i])
    }
    console.log(dp) // [ -2, 1, -2, 4, 3, 5, 6, 1, 5 ]
    return max
  }
}

const list = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
const maxSubArray1 = new MaxSubArray([...list])
console.log(maxSubArray1.maxSubArray1()) // 6
const maxSubArray2 = new MaxSubArray([...list])
console.log(maxSubArray2.maxSubArray2()) // 6

File: LeetCode/53MaximumSubarry/MaximumSubarray.java
// 时间复杂度O(n)，空间复杂度O(1)
public class MaximumSubarry {
    // https://soulmachine.gitbooks.io/algorithm-essentials/java/dp/maximum-subarray.html
    public int maxSubArray1(int[] nums) {
        int maxLocal = nums[0];
        int global = nums[0];
        for (int i = 1; i < nums.length; ++i) {
            maxLocal = Math.max(nums[i], nums[i] + maxLocal);
            global = Math.max(global, maxLocal);
        }
        return global;
    }

    // 地址：https://leetcode.com/problems/maximum-subarray/discuss/20193/DP-solution-and-some-thoughts
    public int maxSubArray2(int[] A) {
        int n = A.length;
        int[] dp = new int[n]; // dp[i] means the maximum subarray ending with A[i]
        dp[0] = A[0];
        int max = dp[0];

        for (int i = 1; i < n; i++) {
            dp[i] = A[i] + (dp[i - 1] > 0 ? dp[i - 1] : 0);
            max = Math.max(max, dp[i]);
        }

        return max;
    }
}